pkg/nfs/controllerserver.go

```diff
@@ -307,6 +315,58 @@ func (cs *ControllerServer) internalUnmount(ctx context.Context, vol *nfsVolume)
 	return err
 }
 
+func (cs *ControllerServer) copyFromVolume(ctx context.Context, req *csi.CreateVolumeRequest, dstVol *nfsVolume) error {
+	srcVol, err := getNfsVolFromID(req.GetVolumeContentSource().GetVolume().GetVolumeId())
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	srcPath := getInternalVolumePath(cs.Driver.workingMountDir, srcVol)
+	dstPath := getInternalVolumePath(cs.Driver.workingMountDir, dstVol)
+	klog.V(2).Infof("copy volume from volume %v -> %v", srcPath, dstPath)
+
+	var volCap *csi.VolumeCapability
+	if len(req.GetVolumeCapabilities()) > 0 {
+		volCap = req.GetVolumeCapabilities()[0]
+	}
+	if err = cs.internalMount(ctx, srcVol, nil, volCap); err != nil {
+		return status.Errorf(codes.Internal, "failed to mount src nfs server: %v", err.Error())
+	}
+	defer func() {
+		if err = cs.internalUnmount(ctx, srcVol); err != nil {
+			klog.Warningf("failed to unmount nfs server: %v", err.Error())
+		}
+	}()
+	if err = cs.internalMount(ctx, dstVol, nil, volCap); err != nil {
+		return status.Errorf(codes.Internal, "failed to mount dst nfs server: %v", err.Error())
+	}
+	defer func() {
+		if err = cs.internalUnmount(ctx, dstVol); err != nil {
+			klog.Warningf("failed to unmount dst nfs server: %v", err.Error())
+		}
+	}()
+
+	// recursive 'cp' with '-a' to handle symlinks. Note that the source path must include trailing '/.',
+	// which is the reason why 'filepath.Join()' is not used as it would perform path cleaning
+	out, err := exec.Command("cp", "-a", fmt.Sprintf("%v%v.", srcPath, filepath.Separator), dstPath).CombinedOutput()
```

Comment:
per the sanity test failure,
```
controllerserver.go:351] copied /tmp/sanity-controller-source-vol-1BFA11DF-89190C95/sanity-controller-source-vol-1BFA11DF-89190C95 -> /tmp/sanity-controller-vol-from-vol-1BFA11DF-89190C95/sanity-controller-vol-from-vol-1BFA11DF-89190C95 output: cp: cannot stat '/tmp/sanity-controller-source-vol-1BFA11DF-89190C95/sanity-controller-source-vol-1BFA11DF-89190C9547.': No such file or directory
```
it should be `cp -a /tmp/sanity-controller-source-vol-1BFA11DF-89190C95/* /tmp/sanity-controller-vol-from-vol-1BFA11DF-89190C95/` ?

Reply:
I think it is expected given the implementation of `getInternalVolumePath()` because volume has a subdir
```
I0314 09:08:26.078270    7781 utils.go:95] GRPC response: {"volume":{"volume_context":{"server":"127.0.0.1","share":"/","subdir":"sanity-controller-source-vol-1BFA11DF-89190C95"},"volume_id":"127.0.0.1##sanity-controller-source-vol-1BFA11DF-89190C95#"}}
```
https://github.com/kubernetes-csi/csi-driver-nfs/blob/af29ce4d9ff5396cc22cb6c1e18bddcfab23338f/pkg/nfs/controllerserver.go#L367-L374

but not sure why the dir doesn't exist and also pretty interesting that `src` and `dst` are the same, probably will need to implement a heuristic for such case because naive usage of `cp` will fail
```
$ cp -a /tmp/a/. /tmp/a
cp: '/tmp/a/.' and '/tmp/a/.' are the same file

$ echo $?
1
```

Reply:
oh, I see it, it's like filepath separator didn't get to the output for some reason
```
ACTUAL SRC PATH: /tmp/sanity-controller-source-vol-1BFA11DF-89190C95/sanity-controller-source-vol-1BFA11DF-89190C9547.
EXPECT SRC PATH: /tmp/sanity-controller-source-vol-1BFA11DF-89190C95/sanity-controller-source-vol-1BFA11DF-89190C9547/.
```

Reply:
tests are passing now, using `/` instead of `filepath.Separator` did the trick